scripts/aws_replicate.py

```diff
@@ -0,0 +1,376 @@
+        )
+
+
+class AWSBucketReplication(object):
+    def __init__(self, global_config, manifest_file, thread_num, job_name, bucket=None):
+        """
+        Class constructor
+
+        Args:
+            global_config(dict): configuration dictionary
+            manifest_file(str): manifest file
+            thread_num(int): number of threads
+            job_name(str): copying|indexing
+            bucket(str): source bucket
+        
+        """
+        self.bucket = bucket
+        self.thread_list = []
+        self.manifest_file = manifest_file
+        self.global_config = global_config
+        self.thread_list = []
+        self.thread_num = thread_num
+        self.json_log = {}
+        self.job_name = job_name
+
+        copied_objects = self.get_copied_objects()
+        for i in xrange(0, self.thread_num):
+            thread = singleThread(
+                str(i),
+                "thread_{}".format(i),
+                self.bucket,
+                copied_objects,
+                self.global_config,
+                self.job_name,
+                workQueue,
+            )
+            self.thread_list.append(thread)
+
+    def get_copied_objects(self):
+        """
+        get all copied objects so that we don't have to re-copy them
+        """
+        s3 = boto3.resource("s3")
+        existed_keys = set()
+        for _, value in PROJECT_MAP.iteritems():
+            for label in ["-open", "-controlled"]:
+                bucket_name = "gdc-" + value + label
+                bucket = s3.Bucket(bucket_name)
+                try:
+                    for file in bucket.objects.all():
+                        existed_keys.add(bucket_name + "/" + file.key)
+                except Exception:
+                    pass
+        return existed_keys
+
+    def prepare(self):
+        """
+        Read data file info from manifest and organize them into  groups.
+        Each group contains files should be copied to same bucket
+        The groups will be push to the queue
+        """
+        if self.manifest_file.startswith("s3://"):
+            submitting_files, _ = get_fileinfo_list_from_s3_manifest(self.manifest_file)
+        else:
+            submitting_files, _ = get_fileinfo_list_from_manifest(self.manifest_file)
+
+        file_grps = exec_files_grouping(submitting_files)
+        mutexLock.acquire()
+        for _, files in file_grps.iteritems():
+            chunk_size = self.global_config.get("chunk_size", 1)
+            idx = 0
+            while idx < len(files):
+                workQueue.put(files[idx : idx + chunk_size])
+                idx = idx + chunk_size
+        mutexLock.release()
+
+        for th in self.thread_list:
+            th.start()
+
+    def run(self,):
+        global TOTAL_PROCESSED_FILES
+
+        start = timeit.default_timer()
+        # Wait for queue to empty
+        while not workQueue.empty():
+            time.sleep(5)
```

Comment:
I think a `threading.Event` is likely to be more elegant; you can have a thread to watch for completion, and a thread to print timing/progress